pkg/terraform/exec/plugins/plugin_versions_generate.go

```diff
@@ -0,0 +1,249 @@
+	re := regexp.MustCompile(`^[[:alnum:]][[:alnum:]\.\-/]+(terraform-provider-([[:alnum:]]+))`)
+
+	// Find all terraform providers
+	for _, r := range file.Require {
+		matches := re.FindStringSubmatch(r.Mod.Path)
+		if len(matches) == 3 {
+			(*plugins)[matches[1]] = &TFPluginInfo{
+				Module: r.Mod,
+				Name:   matches[2],
+			}
+		}
+	}
+
+	// Update replaced terraform providers
+	for _, r := range file.Replace {
+		matches := re.FindStringSubmatch(r.Old.Path)
+		if len(matches) == 3 {
+			(*plugins)[matches[1]] = &TFPluginInfo{
+				Module: r.New,
+				Name:   matches[2],
+			}
+		}
+	}
+}
+
+// getTerraformPluginVersions gathers the terraform provider version numbers
+func getTerraformPluginVersions(file *modfile.File, plugins *map[string]*TFPluginInfo) {
+
+	// Output regex - clean up git name-rev output
+	ore := regexp.MustCompile(`[[:space:]]+|\^[[:digit:]]+`)
+
+	// Tag regex - match the git tag and hash
+	tre := regexp.MustCompile(`[[:digit:]]{14}-([a-fA-F0-9]{12})$`)
+
+	// URL regex - chop off anything after the terraform provider name
+	ure := regexp.MustCompile(`^(.+terraform-provider-[[:alnum:]]+)/v[[:digit:]]+`)
+
+	// Version regex - chop off anything that is not the version
+	vre := regexp.MustCompile(`([0-9\.]+)`)
+
+	for p, pi := range *plugins {
+		var repo *git.Repository
+		var mod *module.Version = &pi.Module
+
+		// XXX: we assume https checkouts
+		url := fmt.Sprintf("https://%s", mod.Path)
+
+		// If this is a v2 or up  module, chop off the end
+		matches := ure.FindStringSubmatch(mod.Path)
+		if len(matches) == 2 {
+			url = fmt.Sprintf("https://%s", matches[1])
+		}
+
+		// git clone if directory doesn't exist, else git pull
+		tmpDir := filepath.Join("tmp", p)
+		if _, err := os.Stat(tmpDir); os.IsNotExist(err) {
+			os.MkdirAll(tmpDir, 0700)
+			repo, err = git.PlainClone(tmpDir, false, &git.CloneOptions{
+				URL:      url,
+				Progress: os.Stdout,
+			})
+			if err != nil {
+				log.Fatal("failed to clone repo ", url)
+			}
+		} else {
+			repo, err = git.PlainOpen(tmpDir)
+			if err != nil {
+				log.Fatal("failed to open repo ", tmpDir)
+			}
+			worktree, err := repo.Worktree()
+			if err != nil {
+				log.Fatal(err)
+			}
+			err = worktree.Pull(&git.PullOptions{})
+			if err != nil && err != git.NoErrAlreadyUpToDate {
+				log.Fatal(err)
+			}
+		}
+
+		pi.Tag = mod.Version
+
+		// If the version is a pseudo version, get the commit hash and
+		// resolve it to a tag if one exists, else default to v1.0.0
+		matches = tre.FindStringSubmatch(mod.Version)
+		if len(matches) == 2 {
+			h := matches[1]
+			hash, err := repo.ResolveRevision(plumbing.Revision(h))
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			here, err := os.Getwd()
+			if err != nil {
+				log.Fatal(err)
+			}
+			err = os.Chdir(tmpDir)
+			if err != nil {
+				log.Fatal(err)
+			}
+			// XXX: HACK: fork git process until go-git supports name-rev
+			out, err := exec.Command("git", "name-rev", "--tags", "--name-only", hash.String()).Output()
```

Comment:
All that happens here is anything extra after the tag, such as ^2 or ~3 is chopped off. The tag is saved and the versions extracted from it.

Reply:
Right. I get that. My question is more around how we are determining what the correct version is, when it could be a tag from an earlier commit or a later commit.